doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,405 @@
+if the application does not currently have a console handle.
+
+> **Note**
+> **Backwards Compatibility**: The behavior of `AllocConsole()` is not changing in response to this specification;
+> therefore, applications that intend to run on older versions of Windows that do not support console allocation
+> policies, which call `AllocConsole()`, will continue to behave normally.
+
+### New APIs
+
+Because a console-subsystem application may still want fine-grained control over when and how its console window is
+spawned, we propose the inclusion of a new API, `AllocConsoleEx(PCONSOLE_ALLOCATE_INFO)`.
+
+#### `AllocConsoleEx`
+
+```c++
+#define CONSOLE_ALLOCATE_NORMAL         0x00000001
+#define CONSOLE_ALLOCATE_HIDDEN         0x00000002
+#define CONSOLE_ALLOCATE_USE_SHOWWINDOW 0x00000004
+
+typedef
+struct _CONSOLE_ALLOCATE_INFO
+{
+    DWORD cbSize;
+    DWORD dwFlags;
+    WORD  wShowWindow;
+    COORD dwWindowSize;
+} CONSOLE_ALLOCATE_INFO, *PCONSOLE_ALLOCATE_INFO;
+
+WINBASEAPI
+HRESULT
+WINAPI
+AllocConsoleEx(PCONSOLE_ALLOCATE_INFO pAllocateInfo);
+```
+
+**AllocConsoleEx** affords an application control over how its console windows are allocated.
+
+> [!NOTE]
+> Unlike `AllocConsole`, which always results in a console window being created, `AllocConsoleEx` without flags will
+> only allocate a console if one was requested during `CreateProcess`.
+>
+> To override this behavior, pass one of `CONSOLE_ALLOCATE_NORMAL` (which is equivalent to being spawned with
+> `CREATE_NEW_WINDOW`) or `CONSOLE_ALLOCATE_HIDDEN` (which is equivalent to being spawned with `CREATE_NO_WINDOW`.)
+
+> [!IMPORTANT]
+> _Why does this API seem so complicated?_
+>
+> Imagine PowerShell switches to `consoleAllocationPolicy` `detached`. Because this new policy
+> [overrides `CREATE_NEW_WINDOW`](#interaction-with-existing-apis), PowerShell has no way of knowing whether the
+> application that launched it truly desired a console window (either visible or hidden.)
+>
+> If PowerShell calls `AllocConsole()`, it will receive a new console window _even if it was spawned with
+> `DETACHED_PROCESS`_. This would result in a significant user experience regression.
+
+##### Parameters
+
+**pAllocateInfo**: A pointer to a `CONSOLE_ALLOCATE_INFO`.
+
+##### `CONSOLE_ALLOCATE_INFO`
+
+###### Members
+
+**cbSize**: Must be set to `sizeof(CONSOLE_ALLOCATE_INFO)`.
+
+**dwFlags**: See the table below for the descriptions of the available flags.
+
+**wShowWindow**: If `CONSOLE_ALLOCATE_USE_SHOWWINDOW` is set, specifies the ["show command"] used to display your
+console window.
+
+**dwWindowSize**: Specifies the viewport size of your newly-allocated console window. If you do not specify a size (e.g.
+you leave it set to `COORD{ 0, 0 }`), your process' console window will be created with a size in accordance with system
+configuration and/or the values in the `STARTUPINFO` structure used during your process' creation.

```

Comment:
We should add a note that says that the size is advisory only.

Reply:
We canned the entire thing, since it turns out that it's _not actually possible_